batchproducer/batchproducer.go

```diff
@@ -233,15 +232,22 @@ func (b *batchProducer) Stop() error {
 }
 
 func (b *batchProducer) setRunning(running bool) {
-	b.runningMu.Lock()
-	defer b.runningMu.Unlock()
-	b.running = running
+	var newValue int32
+	if running {
+		newValue = 1
+	} else {
+		newValue = 0
+	}
+
+	oldValue := b.running
+
+	for swapped := false; !swapped; {
+		swapped = atomic.CompareAndSwapInt32(&b.running, oldValue, newValue)
+	}
 }
 
 func (b *batchProducer) isRunning() bool {
-	b.runningMu.Lock()
-	defer b.runningMu.Unlock()
-	return b.running
+	return b.running == 1
```

Comment:
`atomic.LoadInt32(&b.running) == 1`

Reading an `int32` in any architecture that can fit a 32bit word _should_ be an atomic operation but I have no clue what kind of thread-local optimizations golang does (and they don't have the `volatile` keyword) so it's always safer to mirror atomic writes with atomic reads (the `Load*()` functions.)